Rework print module

sszajbely opened the pull request:
- Split up `PrintController` into `PrinterController`, `PrintJobController`, `PrintAccountController`, `PrintAccountHistoryController`, `FreePagesController`.
- Tried to use REST standards wherever it was possible.
- Removed `Printer` utility class, and printing related commands from the `Commands` utility.
- Removed printer role
- Fixed some bugs regarding paginated tables.
- Added `printers` table. This is mostly done for better REST compatibility and cleaner code, but still is useful if (when) multiple printers are installed at some point.
- Documents are now deleted after they are sent to the printer.

Closes #51, #374, #366 and #344

sszajbely committed:
refactor

sszajbely committed:
remove printer role

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in d330b8b according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely committed:
Merge remote-tracking branch 'origin/development' into print-module-refactor

sszajbely committed:
remove printer role

sszajbely committed:
Merge branch 'print-module-refactor' of github.com:EotvosCollegium/mars into print-module-refactor

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 74b27a2 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely committed:
solving todos

sszajbely committed:
replace exec with process->run

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in c8f89bd according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely committed:
added missing doc strings

sszajbely committed:
Merge branch 'print-module-refactor' of github.com:EotvosCollegium/mars into print-module-refactor

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 6612ad8 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely committed:
add missing doc comments

sszajbely committed:
Merge branch 'print-module-refactor' of github.com:EotvosCollegium/mars into print-module-refactor

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 26c746f according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely committed:
fix

sszajbely committed:
Merge branch 'print-module-refactor' of github.com:EotvosCollegium/mars into print-module-refactor

sszajbely committed:
add missing doc comment

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 62252a9 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely committed:
fixes and #344

sszajbely committed:
Merge branch 'print-module-refactor' of github.com:EotvosCollegium/mars into print-module-refactor

sszajbely commented:
I still have to write the tests, but the main refactoring is done.

horcsinbalint commented on .vscode/settings.json:
I think this file could potentially be removed from git and added to .gitignore

sszajbely replied:
I don't even know why that's there. Good catch

kdmnk commented on app/Console/Commands/CreateDefaultPrinter.php:
```diff
@@ -0,0 +1,39 @@
+use App\Models\Printer;
+use Illuminate\Console\Command;
+
+class CreateDefaultPrinter extends Command
```
Why do you need this? If needed, then let's make this general so we can provide the routers attributes in command arguments.

sszajbely replied:
This is basically an initialization command, so that inserting a printer into the db by hand is not needed. I can add customization if you think it's necessary.

kdmnk replied:
But when do you call it? Can't you just create it in a migration then?

kdmnk commented on app/Http/Controllers/Dormitory/Printing/FreePagesController.php:
```diff
@@ -0,0 +1,83 @@
+     * @param null|string $filter Decides wether all `FreePages` or just the user's `FreePages` should be listed.
+     * @return LengthAwarePaginator
+     */
+    public function indexFreePages(?string $filter = null)
```
Call this only 'index' and do not handle the not admin part here. Modify the table in the view to use data passed with the normal print index page's data. That's what I did with the mac addresses, you can check. I know it's too simple but we should follow that here then.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+    {
+        $request->validate([
+            'amount' => 'required|integer',
+            'user' => 'required|exists:users,id', // Normally this would be a path parameter for the PrintAccount, but we can't do that because of the limitations of blade templates
```
Why? For the select input, you should be able to pass the printaccount id's as keys.

sszajbely replied:
I think the main point is that it would be a path param, not a request param. But still if you can tell me how I can do that in the blade, that would be nice.

kdmnk replied:
Ah, I see. I have an example how we handle that in the tabulator, with javascript: 
```
.ajax({
                    type: "POST",
                    url: "{{ route('internet.internet_accesses.extend', [':id']) }}".replace(':id', id),
                    ...})
```
You need some js to override the default form sending logic and replace the URL, but I think it might worth it here

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+        $printAccount = User::find($request->get('user'))->printAccount;
+
+        // If user can not even transfer balance, we can stop here
+        if (user()->cannot('transferBalance', $printAccount)) {
```
Why not use $this->authorize? (everywhere)

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+
+            $printAccount->update([
+                'balance' => $printAccount->balance - $amount,
+                'last_modified_by' => user()->id,
```
Does it make sense to store the last modified by? We should send an email every time with the actors stated. That should be enough, shouldn't it? Then you could just use increment/decrement. Or if we keep it, move that setter to an observer.

sszajbely replied:
I pretty much kept the functionality of the past code, so this is what was here before, just refactored.

kdmnk replied:
Then, if you want, open a new issue.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+            return redirect()->back()->with('message', __('general.successful_transaction'));
+        }
+        // This is a modification of the current account
+        else {
```
Nah, I would rather separate the endpoints here.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountHistoryController.php:
```diff
@@ -0,0 +1,34 @@
+     * @throws BindingResolutionException
+     * @throws InvalidArgumentException
+     */
+    public function indexPrintAccountHistory()
```
Call this index

kdmnk replied:
@machiato32

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+     * @param null|string $filter Decides wether all `PrintJob`s or just the user's `PrintJob`s should be listed.
+     * @return LengthAwarePaginator
+     */
+    public function indexPrintJobs(?string $filter = null)
```
Call this index

kdmnk replied:
@machiato32

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+     * @throws BindingResolutionException
+     * @throws InvalidArgumentException
+     */
+    private function paginatorFrom(Builder $printJobs, array $columns)
```
If we always use this logic, why don't you modify the api for the tabulatorpaginator?

sszajbely replied:
Because I didn't know if it is used everywhere or not :D

kdmnk replied:
We only use it in AdminInternetController and  MacAddressController elsewhere. Check them please, or you can open a issue.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+     */
+    public function indexPrintJobs(?string $filter = null)
+    {
+        if ($filter === "all") {
```
Split this into 2 functions or create a new AdminPrintJobController

kdmnk replied:
This is still valid. Like what you did in FreePagesController to keep it consistent.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+     */
+    public function store(Request $request)
+    {
+        xdebug_break();
```
Delete

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+    public function store(Request $request)
+    {
+        xdebug_break();
+        Log::info($request->all());
```
Delete

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+            'printer_id' => 'exists:printers,id',
+            'use_free_pages' => 'in:on,off',
+        ]);
+        Log::info("Printjob cost 2");
```
Delete...

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+        $file = $request->file('file');
+
+        /** @var Printer */
+        $printer = $request->has('printer_id') ? Printer::find($request->input("printer_id")) : Printer::firstWhere('name', config('print.printer_name'));
```
Why are there 2 different ways to provide the printer? Can't you move the ID to the url?

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+        ]);
+        Log::info("Printjob cost 2");
+
+        $useFreePages = $request->boolean('use_free_pages');
```
Only keep this variables if you use them multiple times (I can't see now from mobile :D)

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+        }
+
+        $jobId = null;
+        try {
```
Can't you move the try catch inside the function? 

$jobId = $printer->print...
Storage::delete($path)
if (jobId == null) return back()...

kdmnk replied:
Ah I see, I'm stupid, sorry

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrinterController.php:
```diff
@@ -0,0 +1,61 @@
+     */
+    public function index(?string $page = null)
+    {
+        if ($page === "admin") {
```
Create AdminPrintController or split it into two functions

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+            $amount = $request->get('amount');
+
+            // This would be effectively stealing printing money from the other account
+            if ($amount < 0) {
```
move this to validation

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+        // This is a transfer between accounts
+        if ($otherAccount !== null) {
+            // Cannot transfer to yourself
+            if ($otherAccount->user_id === $printAccount->user_id) {
```
why not? keep this simple

horcsinbalint replied:
Is $otherAccount and $printAccount the same PHP object? I'm not sure. If they are different, decreasing $printAccount's balance and saving it, and then increasing $otherAccount's balance and saving it might add money to the account in the end.
Like:
```
a = 1500
b = a #1500
c = a #1500
amount = 300 
b -= amount #1200
a = b #1200
c += amount #1800
a = c #1800
```

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+            }
+
+            // Cannot transfer from other user's account (even if you are admin)
+            if ($printAccount->user_id !== user()->id) {
```
Don't need this if you separate the endpoint (see below)

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+            $amount = $request->get('amount');
+
+            if ($amount < 0 && $printAccount->balance < $amount) {
+                $this->returnNoBalance();
```
are you sure this will return from the function here?

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+            (!$useFreePages && $printAccount->hasEnoughBalance($pageNumber, $copyNumber, $twoSided)))
+        ) {
+            return back()->with('error', __('print.no_balance'));
+        }
```
Can you move both cases into the hasEnoughBalance? (and pass the useFreePages var). Or just make the condition clearer, it's hard to see what is happening with all the negations. Or move this to custom validation rules.

kdmnk replied:
Also, you are calling getFreePagesNeeded/getBalanceNeeded later, which seems like a similar logic.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+            PrintAccount::getFreePagesNeeeded($pageNumber, $copyNumber, $twoSided) :
+            PrintAccount::getBalanceNeeded($pageNumber, $copyNumber, $twoSided);
+
+        Log::info("Printjob cost: $cost");
```
Make one log message that has all the relevant info, if you need. A printjob cost entry on its own does not say much.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+
+            // Remove value regarding the free page checkbox from the session
+            session()->remove('use_free_pages');
+        }
```
This logic should be elsewhere. Create unit tests for this, if you don't have yet.

kdmnk replied:
Still relevant. Modifying free pages or balance should be at least in a helper function. Also the document should be printed after we modified the balance as it can not be rolled back on fail.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+     * @param PrintJob $job
+     * @return RedirectResponse
+     */
+    public function update(PrintJob $job)
```
Then name it cancel. Or if you want to stick with restful names, in terms of print jobs, this should be 'destroy' with a soft delete, which makes sense to me.

kdmnk replied:
Or name it update but add another parameter what should be the new status. Maybe later we can reimburse prints or whatever.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+
+        $jobId = null;
+        try {
+            $jobId = $printer->print($twoSided, $copyNumber, $path);
```
Let's call this 'startPrintJob', if this is supposed to be that,

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+        $this->authorize('update', $job);
+
+        if ($job->state === PrintJobStatus::QUEUED) {
+            $result = ($job->printer ?? Printer::firstWhere('name', config('print.printer_name')))->cancelPrintJob($job);
```
same, why the printer is not required for a job?

sszajbely replied:
Mostly because I didn't want to always have the printer as a dependency for the blade, but I guess it's not really worth much.

kdmnk commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,231 @@
+                        ]);
+                    } else {
+                        $printAccount->balance += $job->cost;
+                    }
```
This should happen automatically, through observers. If you use soft delete, you can also use the restore event here. And make a function for the printAcount to add/subtract free pages from/to the right free pages model. You should not worry about the free pages representation here.

kdmnk commented on app/Models/PrintAccount.php:
```diff
@@ -54,31 +57,101 @@ class PrintAccount extends Model
+     * The free pages which are currently available. Sorts the free pages by their deadline.
+     * @return Collection
+     */
+    public function getAvailableFreePagesAttribute()
```
Use Laravel's Attribute class. You can see an example in MacAddress. Or, rather make this a more constrained hasMany relation getter without the '->get()' at the end.

kdmnk commented on app/Models/PrintAccount.php:
```diff
@@ -54,31 +57,101 @@ class PrintAccount extends Model
+     * @param mixed $twoSided
+     * @return int|float
+     */
+    public static function getFreePagesNeeeded(int $pages, $copies, $twoSided)
```
call this 'getPagesNeeded'

kdmnk commented on app/Models/PrintJob.php:
```diff
@@ -37,46 +41,79 @@ class PrintJob extends Model
+    /**
+     * Attribute for the translated cost.
+     */
+    public function getTranslatedCostAttribute()
```
Use Attributes. (See MacAddress)

kdmnk commented on app/Models/PrintJob.php:
```diff
@@ -37,46 +41,79 @@ class PrintJob extends Model
+    /**
+     * Attribute for the translated state.
+     */
+    public function getTranslatedStateAttribute()
```
Same

kdmnk commented on app/Models/PrintJob.php:
```diff
@@ -37,46 +41,79 @@ class PrintJob extends Model
+    public static function checkAndUpdateStatuses()
+    {
+        foreach(PrintJob::query()->where('state', PrintJobStatus::QUEUED)->whereNotNull('printer_id')->pluck('printer_id')->unique() as $printer_id) {
+            $printJobs = Printer::find($printer_id)->getCompletedPrintJobs();
```
Can't you get failed jobs or warnings also?

kdmnk commented on app/Models/Printer.php:
```diff
@@ -0,0 +1,154 @@
+        if ($result['exit_code'] == 0) {
+            return PrinterCancelResult::Success;
+        }
+        if (strpos($result['output'], "already canceled") !== false) {
```
isn't this a typo?

kdmnk commented on app/Models/Printer.php:
```diff
@@ -0,0 +1,154 @@
+            $jobId = intval($matches[1]);
+        } catch (\Exception $e) {
+            Log::error("Printing error at line: " . __FILE__ . ":" . __LINE__ . " (in function " . __FUNCTION__ . "). " . $e->getMessage());
+            throw new PrinterException($e->getMessage(), $e->getCode(), $e->getPrevious());
```
I see now. Just return null or -1, the new exception with the same data won't give us value.

kdmnk commented on app/Models/Printer.php:
```diff
@@ -0,0 +1,154 @@
+        if (config('app.debug')) {
+            return -1;
+        }
+        $jobId = null;
```
move the return statement inside the try-catch and remove this.

kdmnk commented on app/Models/PrintAccount.php:
```diff
@@ -54,31 +57,101 @@ class PrintAccount extends Model
+    {
+        $pageTypesNeeded = self::getPageTypesNeeded($pages, $twoSided);
+
+        return $pageTypesNeeded['one_sided'] * config('print.one_sided_cost') * $copies +
```
Pretty nice, but do we really want to charge separately for the last page in a two-sided document? Without that, the getPageTypesNeeded would be unnecessary and even the Printer::getDocumentPageNumber could just return the actual pages to be printed in total, directly, based on $twoSided. It would really simplify things without barely anyone noticing or complaining.

sszajbely replied:
I don't know, if you think it's worth it being simpler, I can live with that. :D

kdmnk commented on app/Models/Printer.php:
```diff
@@ -0,0 +1,154 @@
+     */
+    public function getCompletedPrintJobs() {
+        try {
+            $process = new Process(['lpstat', '-W', 'completed', '-o', $this->name, '-h', "$this->ip:$this->port", '|', 'awk', "'{print $1}'"]);
```
Let's create our own Process class that inherits this. Add a parameter to 'getOutput' with a default value returned when we are in debug mode. Override run() accordingly.

kdmnk commented on app/Models/Printer.php:
```diff
@@ -0,0 +1,154 @@
+     * @param string $path 
+     * @return int 
+     */
+    public static function getDocumentPageNumber(string $path): int {
```
This should be elsewhere. Does it make sense to create a Document class, storing the path, cost and other attributes?

kdmnk replied:
Move getBalanceNeeded and similar functions also there.

sszajbely replied:
You mean a utility class, not a model, right?

kdmnk commented on app/Models/Printer.php:
```diff
@@ -0,0 +1,154 @@
+     * @param PrintJob $printJob 
+     * @return PrinterCancelResult
+     */
+    public function cancelPrintJob(PrintJob $printJob) {
```
this should be in the printJob class

sszajbely replied:
The `PrintJob` should have a `cancel()` function, which calls this function. As the `Printer` is the one cancelling the `PrintJob`, but the `PrintJob` is being cancelled. Right?

kdmnk commented on app/Models/Printer.php:
```diff
@@ -0,0 +1,154 @@
+     * @return int The `jobId` belonging to the printjob
+     * @throws PrinterException If the printing fails
+     */
+    public function print(bool $twoSided, int $copies, string $path) {
```
This should be the constructor of the PrintJob class, shouldn't be?

sszajbely replied:
I thought about doing it like that, but I think we would need to alter the db structure a bit. Because a PrintJob not only saves the printjob itself, but also its cost etc. So I think it would be a bit strange to put all of that cost calculation and print account modification into the Printer model aswell.

kdmnk commented on app/Models/PrintJob.php:
```diff
@@ -37,46 +41,79 @@ class PrintJob extends Model
+    /**
+     * Gets the printjob-status with every printer, updates the status of the completed printjobs.
+     */
+    public static function checkAndUpdateStatuses()
```
This should be a method of the Printer, updating it's own printJobs. The loop can be outside the call.

kdmnk reviewed (changes requested):
Nice job, thanks for taking care of this. Please take a look at the issues. Overall, the functions are still a bit too complex, let's revise this again after you fix these. And we need extensive unit tests here (calculating price, pages, etc), otherwise this would be worthless. Let me know if you have questions.

horcsinbalint commented on app/Http/Controllers/Dormitory/Printing/PrintAccountController.php:
```diff
@@ -0,0 +1,128 @@
+     * or by transfering money from one account to the other.
+     */
+    public function update(Request $request)
+    {
```
Shouldn't it be atomic? Like some DB transaction or something like that?

kdmnk commented:
@machiato32 do you have any update on this?

sszajbely commented:
> @machiato32 do you have any update on this?

Not yet, sorry.

kdmnk commented:
Now I had a new idea for the admin/not admin controller separation. 
We merge the two pages into one (originally I did not want to do that but now it suddenly makes sense :D ).
In the backend, we can check based on policies what to query. On the UI we also check the policies if we include the "user" column and other advanced filters for admins or not. 
However, we set a default filter on the current user so no data of other users are shown initially.
The modifier forms can be added next to their respective tables.
What do you think?

sszajbely commented:
> Now I had a new idea for the admin/not admin controller separation. We merge the two pages into one (originally I did not want to do that but now it suddenly makes sense :D ). In the backend, we can check based on policies what to query. On the UI we also check the policies if we include the "user" column and other advanced filters for admins or not. However, we set a default filter on the current user so no data of other users are shown initially. The modifier forms can be added next to their respective tables. What do you think?

I honestly think this seems like more effort than it's worth. I'll just split the controllers and that should be easy to understand for everyone.

sszajbely committed:
Created Process class, created PrinterHelper class, code improvements, fixes.

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in d6fbe58 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely commented:
@kdmnk I tried to resolve everything you mentioned, and also found some bugs.

I still couldn't really test this, and unit/feature tests are also missing, but I think there are a lot of improvements already.

kdmnk commented on app/Policies/PrintAccountPolicy.php:
```diff
@@ -43,4 +40,14 @@ public function modify(User $user): bool
     {
         return false;
     }
+    /**
+     * Determine whether the user can transfer balance from the print account.
+     * @param User $user
+     * @param PrintAccount $printAccount
+     * @return bool
+     */
+    public function transferBalance(User $user, PrintAccount $printAccount): bool
+    {
+        return $user->id == $printAccount->user_id;
+    }
```
yes

kdmnk commented on .vscode/launch.json:
```diff
@@ -0,0 +1,48 @@
+{
```
do we want these files in git?

kdmnk commented on app/Models/PrintJob.php:
```diff
@@ -37,46 +42,92 @@ class PrintJob extends Model
     {
-        return function ($data) {
-            $data->cost = "{$data->cost} HUF";
+        $printer = $this->printer ?? Printer::firstWhere('name', config('print.printer_name'));
```
Again, why don't we always have a printer set in the db? I would delete the config entry or use only that to avoid confusion.

kdmnk commented on routes/web.php:
```diff
@@ -111,22 +114,20 @@
-        Route::post('/print/modify_balance', [PrintController::class, 'modifyBalance'])->name('print.modify');
-    });
-    Route::post('/print/add_free_pages', [PrintController::class, 'addFreePages'])->name('print.free_pages')->middleware('can:create,App\Models\FreePages');
+    Route::get('/printer/{page?}', [PrinterController::class, 'index'])->name('printer.index'); // also print.manage, print
```
what are the stuff in comments?

kdmnk reviewed (changes requested):
Looks much cleaner but some comments still applies. Also take a look at the coderabbit comments which are pretty cool.
We can take a look together if you want.

sszajbely committed:
Resolving comments

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 9499653 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely committed:
Merge remote-tracking branch 'origin/development' into print-module-refactor

sszajbely committed:
Merge branch 'print-module-refactor' of github.com:EotvosCollegium/mars into print-module-refactor

kdmnk reviewed (approved):
The code looks nice now, good job and thanks!
Let's test this.

sszajbely committed:
Minor fixes

sszajbely commented:
I would like to hand over the testing phase of this PR to someone who actually still lives in the building, especially since I cannot test the printing itself, which is the most important part of this. @EotvosCollegium/marscodeowners.

sszajbely committed:
Added doc comments

horcsinbalint commented on app/Console/Commands.php:
```diff
@@ -15,65 +16,15 @@ private static function isDebugMode()
             // This happens too often to log.
-            $command = config('commands.ping') . " " . $router->ip . " -c 1 | grep 'error\|unreachable'";
-            $result = exec($command);
+            $process = new Process(['ping', $router->ip, '-c', '1', '|', 'grep', "'error\|unreachable'"]);
```
I have no experience with `Symfony\Component\Process\Process`, but in other languages similar concepts of processes would not allow multiple commands being piped together.

horcsinbalint commented on app/Console/Commands.php:
```diff
@@ -83,10 +34,9 @@ public static function latexToPdf($path, $outputDir)
-            Log::info($command);
-            $result = exec($command);
-            Log::info($result);
+            $process = new Process(['pdflatex', '-interaction=nonstopmode', '-output-dir', $outputDir, $path]);
```
As far as I remember, the `pdflatex` command did not work inside the container in the case of the php interpreter, that is why `config('command.pdlatex')` was used.

sszajbely committed:
Fixed piping in commands.

sszajbely committed:
Use commands config file

sszajbely committed:
Minor fixes.

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in f85af3d according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

sszajbely committed:
Fix for migration.

sszajbely committed:
Merge branch 'print-module-refactor' of github.com:EotvosCollegium/mars into print-module-refactor

sszajbely commented:
The only test failing now is the `PrintControllerTest` which I haven't updated yet. So I think all major concerns asides from tests are being resolved now, so as I stated earlier, I would like to hand off this part to someone in the building @EotvosCollegium/marscodeowners

sszajbely committed:
Minor fixes

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in d656944 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

horcsinbalint commented on app/Http/Controllers/Dormitory/Printing/PrintJobController.php:
```diff
@@ -0,0 +1,185 @@
+use App\Models\PrintAccount;
+use App\Models\Printer;
+use App\Enums\PrinterCancelResult;
+use App\Models\PrinterHelper;
```
`App\Models\PrinterHelper` does not exists, you probably meant `App\Utils\PrinterHelper`

sszajbely replied:
Good catch, namespace was wrong.

sszajbely committed:
Fixed PrinterHelper namespace

sszajbely committed:
Merge branch 'print-module-refactor' of github.com:EotvosCollegium/mars into print-module-refactor

horcsinbalint commented on resources/views/user/printing.blade.php:
```diff
@@ -11,14 +11,14 @@
                         <li>
                         <span class="new badge
-                            @if($page->available())
+                            @if($page->available)
```
`available` is a function, not an attribute. It should be `available()`

sszajbely replied:
It is an attribute though.
```
protected function available()
    {
        return Attribute::make(
            get: fn () => now()->isBefore($this->deadline)
        );
    }
```

horcsinbalint commented on app/Models/FreePages.php:
```diff
@@ -51,23 +53,73 @@ class FreePages extends Model
+     * Wether the free pages are still available.
+     * @return bool
+     */
+    protected function available()
```
This fucntion is called outside from `printing.blade.php`, should not be protected, it should be public.

sszajbely replied:
AFAIK attributes don't care about the visibility of the function behind it. Virtually every attribute in the project is created in a protected function.

horcsinbalint replied:
For me route `/profile` was crashing with error "App\Models\FreePages::available must return a relationship instance.". The return value type should be included according to the documentation (https://laravel.com/docs/11.x/eloquent-mutators#attribute-casting): "All attribute accessor / mutator methods must declare a return type-hint of `Illuminate\Database\Eloquent\Casts\Attribute`":
```php
protected function available() : Attribute
```

horcsinbalint replied:
With this change, it works fine on my machine

sszajbely replied:
Ah so it needs the type hint, okay. That's why I don't like use the new syntax for attribute creation @kdmnk :P. But thanks, will fix it tomorrow

horcsinbalint replied:
Thanks, there are also other instances where it should be fixed (like `App\Models\PrintJob::translatedCost`).

horcsinbalint commented on database/factories/PrintJobFactory.php:
```diff
@@ -13,8 +14,7 @@ public function definition()
     {
         return [
```
PrintJob requires a `printer_id` attribute to be added during seeding as there is no default value.

horcsinbalint commented on app/Console/Commands.php:
```diff
@@ -10,84 +11,23 @@
-            // This happens too often to log.
-            $command = config('commands.ping') . " " . $router->ip . " -c 1 | grep 'error\|unreachable'";
-            $result = exec($command);
+        if (preg_match('/[a-zA-Z]/', $router->ip)) {
```
Nitpick, as this is already an improvement: could we check if it is in fact an IP address, and if not, then throw an exception? Like using this regex? `^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])$`
Some quite painful shellscript could be injected here even if they do not control any letters. (Like a fork bomb: `:(){ :|:& };:`)

horcsinbalint committed:
Remove references for Models in migrations

Referencing Models in migrations should be avoided as changing the Model might change older migrations.
Migrations are usually the "way of propagating changes you make to your Models (adding a field, deleting a model, etc.) into your database schema" (https://docs.djangoproject.com/en/5.1/topics/migrations/).
Therefore migrations should never ever reference the Models as their main purpose is to have a migration between two different versions of models.
Once a migration is written, they should usually never be edited in the future (unless the language / framework changes the syntax). But functionally a migration should remain the same in its whole life regardless of how the application is developed.

horcsinbalint committed:
Fixes

horcsinbalint committed:
Remove printer role using migration

horcsinbalint committed:
Fix seeders

Seeders could not work with an entirely empty database. It's a problem that was introduced by the recent changes in migrations

horcsinbalint committed:
Added better check for pinging command

horcsinbalint committed:
Added logging for the results of commands

It has been extremely useful several times that we had commands of
successful and failed commands.

horcsinbalint committed:
Added prefix for printing

horcsinbalint commented:
I added a few proposed changes in #669

horcsinbalint commented:
I will decrease the scope of #669 , and merge them and try it out in production on Saturday if no other issues are raised

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in afd2f1f according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/394

horcsinbalint committed:
Make .env.example more understandable

horcsinbalint committed:
Using proper filtering instead of regex for ip

Use filter_var instead of regex. Coderabbit suggested the change. I do
not agree with the reasoning coderabbit gave, it was functional, this is
simply more readable.

horcsinbalint committed:
Added more strict rate limiting

horcsinbalint committed:
Fixing up routes

horcsinbalint committed:
Better emailing for e-mail balances

Added a missing Ft to Hungarian email and also emailing the recipient.

horcsinbalint committed:
Emailing all involved

Send emails to each individual of all of the balance changes. (For
better transparency.)

horcsinbalint committed:
Better commenting

horcsinbalint committed:
Restore composer.lock

horcsinbalint committed:
Bugfixes and better suit our needs

Using the printing filesystem for printing.
In case of free pages we are charging for each side we need to put toner
on. We try to minimize our paper usage by automatically selecting
double-sided printing.

horcsinbalint committed:
Command fix

horcsinbalint committed:
Command fix

horcsinbalint committed:
Fix statuses

horcsinbalint committed:
Adding missing translation

horcsinbalint committed:
Fix tests

horcsinbalint committed:
Fix tests

horcsinbalint committed:
Fix tests

horcsinbalint commented:
I gave up on reviewing all of the changes. I fixed a few bugs.
I am merging this for now, but I will revert the PR if anything goes wrong.

horcsinbalint commented:
There was an issue, reverted, a new PR will be soon opened
